# Modelo de Regresion Logistica - ADN Boca (perfil puro)

Entrenamos una regresion logistica para clasificar si un jugador tiene perfil ADN Boca.
Dataset: `adn_boca_real_features.csv` (810 registros, 21 columnas).

## Por que "perfil puro"

La etiqueta se define como `(rating>=7.0) & (goles+asistencias>=3)`. Por lo tanto, cualquier
feature derivada de goles/asistencias (goles, asistencias, goles_por_partido, contribucion_gol,
promedio_*) es una **reconstruccion directa del umbral** de la etiqueta y satura las probabilidades
(se vio en `scouting_resultado_historico.csv` viejo con valores 0.9999...).

Para evitar ese solapamiento, el modelo usa SOLO features de perfil (edad, pases, posicion,
trayectoria, minutos) que explican quien tiene rating alto entre los que producen goles.
Los goles/asistencias quedan en el CSV como columnas informativas del ranking, no como features.

In [1]:
import os
import sys

sys.path.insert(0, os.path.abspath(os.getcwd()) + '/src')
import train_model
from train_model import (cargar_datos, construir_X, split_por_jugador, BASE_FEATURES)

df = cargar_datos()
print(f'Dataset: {df.shape}')
print(f'Target: 0={sum(df.etiqueta==0)}, 1={sum(df.etiqueta==1)}')
print(f'Features de perfil ({len(BASE_FEATURES)}): {BASE_FEATURES}')
print(f'Posiciones agrupadas: {train_model.POS_COLUMNS}')

Dataset: (810, 21)
Target: 0=653, 1=157
Features de perfil (5): ['pases_precisos', 'edad', 'temporadas_en_dataset', 'partidos_por_temporada', 'perfil_ofensivo']
Posiciones agrupadas: ['pos_Defensor_central', 'pos_Delantero', 'pos_Extremo', 'pos_Lateral', 'pos_Mediocampista_central', 'pos_Mediocampista_ofensivo']


## Entrenamiento

Split por jugador (sin data leakage): un jugador va a test si su ultima temporada >= 2023.
Regresion logistica balanceada con GridSearch sobre C (AUC). Se guardan los pkl
(`modelo_adn_boca.pkl`, `scaler.pkl`, `features_list.pkl`, `position_encoder.pkl`)
y se genera `scouting_resultado_historico.csv` (OOF para train, directo para test).

In [2]:
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, roc_auc_score

train_model.main()
print('\nEntrenamiento completo. Ver metricas en la salida de arriba.')

Train: 666 registros, 263 jugadores
Test:  144 registros, 54 jugadores


=== MODELO PRINCIPAL: LogisticRegression balanceada (perfil puro) ===
Mejor C: 10 | Mejor AUC (CV): 0.934
AUC-ROC test: 0.887
              precision    recall  f1-score   support

   No encaja       0.95      0.86      0.90       123
    ADN Boca       0.47      0.71      0.57        21

    accuracy                           0.84       144
   macro avg       0.71      0.79      0.73       144
weighted avg       0.88      0.84      0.85       144



OOF AUC: 0.930

scouting_resultado.csv guardado (810 filas)

=== MODELO ALTERNATIVO: LogisticRegression L1 (saga) ===


Mejor C: 10 | AUC test: 0.866
              precision    recall  f1-score   support

   No encaja       0.95      0.83      0.89       123
    ADN Boca       0.43      0.76      0.55        21

    accuracy                           0.82       144
   macro avg       0.69      0.80      0.72       144
weighted avg       0.88      0.82      0.84       144

  models/modelo_adn_boca.pkl guardado
  models/scaler.pkl guardado
  models/features_list.pkl guardado
  models/position_encoder.pkl guardado
  models/modelo_logistic_l1.pkl guardado
  models/config.pkl guardado
AUC-ROC test: 0.887

Entrenamiento completo. Ver metricas en la salida de arriba.


## Resultado esperado (reproducible via `python src/train_model.py`)

- AUC-ROC test ~0.89 (jugadores nuevos, nunca vistos).
- Probabilidades distribuidas en (0,1), sin saturacion.
- `modelo_logistic_l1.pkl` (penalty='l1', solver='saga') queda como alternativa documentada.